src/execution/ingestion_sample.cpp

```diff
@@ -0,0 +1,680 @@
+	auto ret = make_uniq<IngestionSample>(sample_count);
+
+	ret->base_reservoir_sample = base_reservoir_sample->Copy();
+	ret->destroyed = destroyed;
+
+	if (!sample_chunk || destroyed) {
+		return unique_ptr_cast<IngestionSample, BlockingSample>(std::move(ret));
+	}
+
+	D_ASSERT(sample_chunk);
+
+	// create a new sample chunk to store new samples
+	auto types = sample_chunk->GetTypes();
+	idx_t new_sample_chunk_size =
+	    for_serialization ? NumActiveSamples() : FIXED_SAMPLE_SIZE * FIXED_SAMPLE_SIZE_MULTIPLIER;
+	// how many values should be copied
+	idx_t values_to_copy = MinValue<idx_t>(NumActiveSamples(), FIXED_SAMPLE_SIZE);
+
+	auto new_sample_chunk = CreateNewSampleChunk(types, new_sample_chunk_size);
+
+	// set up selection vector to copy IngestionSample to ReservoirSample
+	SampleCopyHelper sel = GetSelToCopyData(values_to_copy);
+	ret->actual_sample_indexes = sel.actual_indexes;
+	ret->base_reservoir_sample->reservoir_weights = sel.reservoir_weights;
+
+	if (SamplingState() != SamplingMode::FAST) {
+		ret->base_reservoir_sample->min_weighted_entry_index =
+		    ret->base_reservoir_sample->reservoir_weights.top().second;
+	}
+
+	ret->sample_chunk = std::move(new_sample_chunk);
+	ret->UpdateSampleAppend(*sample_chunk, sel.sel, values_to_copy);
+	D_ASSERT(ret->sample_chunk->size() <= FIXED_SAMPLE_SIZE);
+	ret->Verify();
+	return unique_ptr_cast<IngestionSample, BlockingSample>(std::move(ret));
+}
+
+void IngestionSample::ConvertToSlowSample() {
+	base_reservoir_sample->FillWeights(actual_sample_indexes);
+	actual_sample_indexes.clear();
+}
+
+void IngestionSample::SimpleMerge(IngestionSample &other) {
+	D_ASSERT(GetPriorityQueueSize() == 0);
+	D_ASSERT(other.GetPriorityQueueSize() == 0);
+
+	if (other.actual_sample_indexes.empty() && other.GetTuplesSeen() == 0) {
+		return;
+	}
+
+	if (actual_sample_indexes.empty() && GetTuplesSeen() == 0) {
+		actual_sample_indexes = std::move(other.actual_sample_indexes);
+		base_reservoir_sample->num_entries_seen_total = other.GetTuplesSeen();
+		return;
+	}
+
+	idx_t total_seen = GetTuplesSeen() + other.GetTuplesSeen();
+
+	auto weight_tuples_this = static_cast<double>(GetTuplesSeen()) / static_cast<double>(total_seen);
+	auto weight_tuples_other = static_cast<double>(other.GetTuplesSeen()) / static_cast<double>(total_seen);
+
+	// If weights don't add up to 1, most likely a simple merge occured and no new samples were added.
+	// if that is the case, add the missing weight to the lower weighted sample to adjust.
+	// this is to avoid cases where if you have a 20k row table and add another 20k rows row by row
+	// then eventually the missing weights will add up, and get you a more even distribution
+	if (weight_tuples_this + weight_tuples_other < 1) {
+		weight_tuples_other += 1 - (weight_tuples_other + weight_tuples_this);
+	}
+
+	idx_t keep_from_this = 0;
+	idx_t keep_from_other = 0;
+	if (weight_tuples_this > weight_tuples_other) {
```

Comment:
We should probably random the elements we insert in the merge here to avoid introducing a bias to the sample